### PR TITLE
Modified parser. Added lots of checks. Implemented adding new serials. Updated setup checks.

### PR DESCRIPTION
**Pull request by BobMDu**

### CODE REVIEW

File: data/tables.json [hunk 1]
```diff
@@ -0,0 +1,145 @@
+{
```
Comment (tomkralidis) [inline data/tables.json]: Prefer to make this file YAML which will make it easier to maintain over time.

File: woudc_data_registry/models.py [hunk 2]
```diff
@@ -464,7 +464,7 @@ def __init__(self, ecsv):
                 ecsv.extcsv['DATA_GENERATION']['ScientificAuthority']
 
         self.platform_type = ecsv.extcsv['PLATFORM']['Type']
-        self.platform_id = ecsv.extcsv['PLATFORM']['ID']
+        self.platform_id = str(ecsv.extcsv['PLATFORM']['ID'])
```
Comment (tomkralidis) [inline woudc_data_registry/models.py]: Why are we casting?

File: woudc_data_registry/models.py [hunk 3]
```diff
@@ -473,7 +473,7 @@ def __init__(self, ecsv):
 
         self.instrument_name = ecsv.extcsv['INSTRUMENT']['Name']
         self.instrument_model = ecsv.extcsv['INSTRUMENT']['Model']
-        self.instrument_number = ecsv.extcsv['INSTRUMENT']['Number']
+        self.instrument_number = str(ecsv.extcsv['INSTRUMENT']['Number'])
```
Comment (tomkralidis) [inline woudc_data_registry/models.py]: Why are we casting here too?

File: woudc_data_registry/parser.py [hunk 4]
```diff
@@ -150,10 +150,13 @@ def __init__(self, content):
         for row in reader:
             if len(row) == 1 and row[0].startswith('#'):  # table name
                 table_name = row[0].replace('#', '')
-                if table_name in DOMAINS['metadata_tables'].keys():
-                    found_table = True
-                    LOGGER.debug('Found new table {}'.format(table_name))
-                    self.extcsv[table_name] = {}
+                table_count = 2
+                while table_name in self.extcsv:
+                    table_name = table_name + '_' + str(table_count)
```
Comment (tomkralidis) [inline woudc_data_registry/parser.py]: better as `table_name = '{}_{}'.format(table_name, table_count)`

File: woudc_data_registry/parser.py [hunk 5]
```diff
@@ -165,11 +168,17 @@ def __init__(self, content):
...
-                if table_name in DOMAINS['metadata_tables'].keys():
+                if table_name is not None:
                     self.extcsv[table_name]['_line_num'] = \
                         int(reader.line_num + 1)
                     for idx, val in enumerate(row):
-                        field = self.extcsv[table_name]['_fields'][idx]
+                        try:
+                            field = self.extcsv[table_name]['_fields'][idx]
+                        except IndexError:
+                            msg = ('Rows in table {} have too many '
```
Comment (tomkralidis) [inline woudc_data_registry/parser.py]: Better as:

```python
msg = ('Rows in table {} have too many elements.'.format(
    table_name))
```

File: woudc_data_registry/processing.py [hunk 6]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
-        datasets = self.registry.query_distinct(Dataset.identifier)
+        LOGGER.debug('Validating project')
+        self.projects = self.registry.query_distinct(Project.identifier)
+        if self.data_record.content_class not in self.projects:
+            msg = 'Project {} not found in registry'.format(
+                self.data_record.content_class)
+            LOGGER.error(msg)
+            raise ProcessingError(msg)
+        else:
+            LOGGER.debug("Matched with project:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
            LOGGER.debug('Matched with project: {}'.format(
                self.data_record.content_class))
```

File: woudc_data_registry/processing.py [hunk 7]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
-        if self.data_record.content_category not in datasets:
+        LOGGER.debug('Validating dataset')
+        self.datasets = self.registry.query_distinct(Dataset.identifier)
+        if self.data_record.content_category not in self.datasets:
             msg = 'Dataset {} not found in registry'.format(
                 self.data_record.content_category)
             LOGGER.error(msg)
             raise ProcessingError(msg)
+        else:
+            LOGGER.debug("Matched with dataset:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]:
```python
           LOGGER.debug('Matched with dataset: {}'.format(
                self.data_record.content_category))
```

File: woudc_data_registry/processing.py [hunk 8]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+                self.data_record.content_category))
 
-        LOGGER.debug('Validating project')
-        projects = self.registry.query_distinct(Project.identifier)
-
-        if self.data_record.content_class not in projects:
-            msg = 'Project {} not found in registry'.format(
-                self.data_record.content_class)
+        LOGGER.debug('Validating contributor')
+        self.contributors = self.registry.\
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
        self.contributors = self.registry.query_distinct(
            Contributor.identifier)
```

File: woudc_data_registry/processing.py [hunk 9]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+            query_distinct(Contributor.identifier)
+        file_contributor = self.data_record.data_generation_agency + ':' +\
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
        file_contributor = '{}:{}'.format(
            self.data_record.data_generation_agency,
            self.data_record.content_class)
```

File: woudc_data_registry/processing.py [hunk 10]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+            self.data_record.content_class
+        if file_contributor not in self.contributors:
+            msg = 'Contributor {} not found in registry'.format(
+                file_contributor)
+            LOGGER.error(msg)
+            raise ProcessingError(msg)
+        else:
+            LOGGER.debug("Matched with contributor:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
            LOGGER.debug('Matched with contributor: {}'.format(
                file_contributor))
```

File: woudc_data_registry/processing.py [hunk 11]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+            'identifier': self.data_record.platform_id,
+            'name': self.data_record.platform_name,
+            'country_id': self.data_record.platform_country
+        }
+
+        LOGGER.debug('Validating station id...')
+        results = self.registry.query_multiple_fields(
+            Station, station, ['identifier'])
+        if results:
+            LOGGER.debug("Validated with id:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
            LOGGER.debug('Validated with id: {}'.format(
                self.data_record.platform_id))
```

File: woudc_data_registry/processing.py [hunk 12]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
             raise ProcessingError(msg)
 
-        LOGGER.debug('Validating contributor')
-        contributors = self.registry.query_distinct(
-            Contributor.identifier)
+        LOGGER.debug('Validating station name...')
+        fields = ['identifier', 'name']
+        results = self.registry.query_multiple_fields(Station, station, fields)
+        if results:
+            LOGGER.debug("Validated with name:'{}' for id:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
            LOGGER.debug('Validated with name:'{}' for id: {}'.format(
                self.data_record.platform_name, self.data_record.platform_id))
```

File: woudc_data_registry/processing.py [hunk 13]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+                self.data_record.platform_name, self.data_record.platform_id))
+        else:
+            msg = 'Station name:"{}" did not match data for id:"{}"'.format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
            msg = 'Station name {} did not match data for id {}'.format(
                self.data_record.platform_id, self.data_record.platform_id)
```

File: woudc_data_registry/processing.py [hunk 14]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+            raise ProcessingError(msg)
 
-        if self.data_record.data_generation_agency not in contributors:
-            msg = 'Contributor {} not found in registry'.format(
-                self.data_record.data_generation_agency)
+        LOGGER.debug('Validating station country...')
+        fields = ['identifier', 'country_id']
+        results = self.registry.query_multiple_fields(Station, station, fields)
+        if results:
+            LOGGER.debug("Validated with country:'{}' for id:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: `LOGGER.debug('Validated with country {} for id {}'.format(`

File: woudc_data_registry/processing.py [hunk 15]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+                self.data_record.platform_country,
+                self.data_record.platform_id))
+        else:
+            msg = "Station country:'{}' did not match data for id:'{}'".format(
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: `             msg = "Station country {} did not match data for id {}'.format(`

File: woudc_data_registry/processing.py [hunk 16]
```diff
@@ -178,44 +178,151 @@ def process_data(self, infile, verify_only=False):
...
+        self.instruments = self.registry.query_distinct(Instrument.identifier)
+        instrument_added = False
+        instrument = [self.data_record.instrument_name,
+                      self.data_record.instrument_model,
+                      self.data_record.instrument_number,
+                      self.data_record.platform_id,
+                      self.data_record.content_category]
+        instrument_id = ':'.join(instrument)
+        if instrument_id not in self.instruments:
+            instrument[2] = str(int(instrument[2]))
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Why the double cast here?

File: woudc_data_registry/processing.py [hunk 17]
```diff
@@ -232,6 +339,118 @@ def process_data(self, infile, verify_only=False):
...
+        with open(dataset_tables) as jsonfile:
+            tables = json.load(jsonfile)
+
+        if dataset in tables.keys():
+            curr_dict = tables[dataset]
+            if level in curr_dict.keys():
+                curr_dict = curr_dict[level]
+                if form in curr_dict:
+                    curr_dict = curr_dict[form]
+                    if "1" in curr_dict.keys():
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: 1. single quotes 2. should this be (at some point) an `int` ?

File: woudc_data_registry/processing.py [hunk 18]
```diff
@@ -232,6 +339,118 @@ def process_data(self, infile, verify_only=False):
...
+                elif extra:
+                    msg = 'The following fields should not be in table {}'\
+                          ': {}'.format(table, extra)
+                    LOGGER.error(msg)
+                    raise ProcessingError(msg)
+                else:
+                    LOGGER.debug('All fields in table {} '
+                                 'validated'.format(table))
+            else:
+                msg = 'Could not validate as {} is a required table in a {} '\
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: Better as:

```python
                msg = 'Could not validate ({}  requires {})'.format(table,
                     extcsv['CONTENT']['Category'])
```

File: woudc_data_registry/processing.py (hunk 16 above)
Comment (BobMDu) [reply woudc_data_registry/processing.py]: This is to check for the old format by casting it into a integer then creating a instrument_id with it as a string

**Pull request merged**